openslides/utils/cache_providers.py

```diff
@@ -253,21 +252,48 @@ def zadd_args(change_id: int) -> Generator[Union[int, str], None, None]:
         if user_id is None, the full_data is returned. If user_id is an int, the restricted_data
         for an user is used. 0 is for the anonymous user.
         """
-        # TODO: rewrite with lua to get all elements with one request
+        changed_elements: Dict[str, List[bytes]] = defaultdict(list)
+        deleted_elements: List[str] = []
+        if user_id is None:
+            cache_key = self.get_full_data_cache_key()
+        else:
+            cache_key = self.get_restricted_data_cache_key(user_id)
+
+        # Convert max_change_id to a string. If its negative, use the string '+inf'
+        redis_max_change_id = "+inf" if max_change_id < 0 else str(max_change_id)
         async with self.get_connection() as redis:
-            changed_elements: Dict[str, List[bytes]] = defaultdict(list)
-            deleted_elements: List[str] = []
-            for element_id in await redis.zrangebyscore(self.get_change_id_cache_key(), min=change_id):
-                if element_id.startswith(b'_config'):
-                    continue
-                element_json = await redis.hget(self.get_full_data_cache_key(), element_id)  # Optional[bytes]
-                if element_json is None:
-                    # The element is not in the cache. It has to be deleted.
-                    deleted_elements.append(element_id)
-                else:
-                    collection_string, id = split_element_id(element_id)
-                    changed_elements[collection_string].append(element_json)
-            return changed_elements, deleted_elements
+            # lua script that returns gets all element_ids from change_id_cache_key
+            # and then uses each element_id on full_data or restricted_data.
+            # It returns a list where the odd values are the change_id and the
+            # even values the element as json. The function wait_make_dict creates
+            # a python dict from the returned list.
+            elements: Dict[bytes, Optional[bytes]] = await aioredis.util.wait_make_dict(redis.eval(
```

Comment:
Do you want to use Python's textwrap.dedent() or is this not required?

Reply:
It is not required. Taps and leading white space have to meaning in lua. The only downside is, that it is unneeded bytes that have to send to redis each time. But the correct solution for that (in another pull request) is to use https://redis.io/commands/evalsha